core-plugin/src/main/groovy/com/github/jrubygradle/internal/core/AbstractIvyXmlProxyServer.groovy

```diff
@@ -0,0 +1,180 @@
+package com.github.jrubygradle.internal.core
+
+import com.github.jrubygradle.api.core.ApiException
+import com.github.jrubygradle.api.core.IvyXmlProxyServer
+import com.github.jrubygradle.api.core.RubyGemQueryRestApi
+import com.github.jrubygradle.api.gems.GemInfo
+import com.github.jrubygradle.api.gems.GemVersion
+import com.github.jrubygradle.internal.gems.GemToIvy
+import groovy.transform.CompileStatic
+import groovy.transform.InheritConstructors
+import groovy.transform.Synchronized
+import groovy.util.logging.Slf4j
+import org.ysb33r.grolifant.api.ExclusiveFileAccess
+
+import java.nio.file.Files
+import java.nio.file.Path
+import java.time.Instant
+
+import static com.github.jrubygradle.api.gems.GemVersion.gemVersionFromGradleIvyRequirement
+import static com.github.jrubygradle.internal.core.IvyUtils.revisionsAsHtmlDirectoryListing
+import static java.nio.file.Files.move
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING
+
+/** Base class for implementing a proxy IvyXML server.
+ *
+ * @author Schalk W. Cronjé
+ *
+ * @since 2.0
+ */
+@CompileStatic
+@Slf4j
+abstract class AbstractIvyXmlProxyServer implements IvyXmlProxyServer {
+
+    @InheritConstructors
+    static class NotFound extends Exception {
+    }
+
+    /** Tell the server to refresh dependencies upon a next run.
+     *
+     * @param refresh {@code true} to reload dependencies.
+     */
+    @Override
+    void setRefreshDependencies(boolean refresh) {
+        refreshDependencies = refresh ? 1 : 0
+    }
+
+    /** Get the address of the local proxy.
+     *
+     * @return Local address as a URI.
+     */
+    @Override
+    URI getBindAddress() {
+        "http://localhost:${bindPort}".toURI()
+    }
+
+    /** Returns the cache location for a specific GEM.
+     *
+     * @param group Group associated with GEM.
+     * @param name GEM name.
+     * @param revision GEM revision.
+     * @return Location of {@code ivy.xml} file.
+     */
+    @SuppressWarnings('UnusedMethodParameter')
+    Path ivyFile(String group, String name, String revision) {
+        new File(localCachePath, "${name}/${revision}/ivy.xml").toPath()
+    }
+
+    /** Implementation of a proxy server.
+     *
+     * @param cache Root directory for local Ivy XML cache.
+     * @param serverUri URI of remote Rubygems proxy.
+     * @param group Group that will be associated with the Rubygems proxy.
+     */
+    protected AbstractIvyXmlProxyServer(File cache, URI serverUri, String group) {
+        localCachePath = cache
+        gemToIvy = new GemToIvy(serverUri)
+        api = new DefaultRubyGemRestApi(serverUri)
+        this.group = group
+    }
+
+    @Synchronized
+    @SuppressWarnings('BuilderMethodWithSideEffects')
+    protected void createIvyXml(Path ivyXml, String name, String revision) {
```

Comment:
Handles caching of ivy.xml files even across process boundaries (Multiple GRdle daemons writing to the same caching area).